libbeat/publisher/pipeline/output.go

```diff
@@ -71,56 +91,69 @@ func (w *clientWorker) run() {
 }
 
 func (w *netClientWorker) Close() error {
-	w.closed.Store(true)
+	close(w.done)
+	w.wg.Wait()
 	return w.client.Close()
 }
 
-func (w *netClientWorker) run() {
-	for !w.closed.Load() {
-		reconnectAttempts := 0
-
-		// start initial connect loop from first batch, but return
-		// batch to pipeline for other outputs to catch up while we're trying to connect
-		for batch := range w.qu {
-			batch.Cancelled()
-
-			if w.closed.Load() {
-				logp.Info("Closed connection to %v", w.client)
-				return
-			}
+func (w *netClientWorker) start() {
+	w.wg.Add(1)
+	go func() {
+		defer w.wg.Done()
+		w.run()
+	}()
+}
 
-			if reconnectAttempts > 0 {
-				logp.Info("Attempting to reconnect to %v with %d reconnect attempt(s)", w.client, reconnectAttempts)
-			} else {
-				logp.Info("Connecting to %v", w.client)
-			}
+func (w *netClientWorker) run() {
+	var (
+		connected         bool
+		reconnectAttempts int
+	)
+
+	for {
+		// Prioritize on next loop to close the client.
+		select {
+		case <-w.done:
+			logp.Info("Closed connection to %v", w.client)
+			return
+		default:
+		}
 
-			err := w.client.Connect()
-			if err != nil {
-				logp.Err("Failed to connect to %v: %v", w.client, err)
-				reconnectAttempts++
+		// Either we are closing or we are waiting on events.
+		select {
+		case <-w.done:
+			logp.Info("Closed connection to %v", w.client)
+			return
+		case batch := <-w.qu:
+			if batch == nil {
 				continue
 			}
 
-			logp.Info("Connection to %v established", w.client)
-			reconnectAttempts = 0
-			break
-		}
+			if !connected {
+				batch.Cancelled()
 
-		// send loop
-		for batch := range w.qu {
-			if w.closed.Load() {
-				if batch != nil {
-					batch.Cancelled()
+				if reconnectAttempts > 0 {
+					logp.Info("Attempting to reconnect to %v with %d reconnect attempt(s)", w.client, reconnectAttempts)
+				} else {
+					logp.Info("Connecting to %v", w.client)
 				}
-				return
+
+				if err := w.client.Connect(); err != nil {
+					logp.Err("Failed to connect to %v: %v", w.client, err)
+					reconnectAttempts++
+				}
+
+				logp.Info("Connection to %v established", w.client)
+				connected = true
+				reconnectAttempts = 0
+				continue
 			}
 
 			err := w.client.Publish(batch)
 			if err != nil {
 				logp.Err("Failed to publish events: %v", err)
 				// on error return to connect loop
-				break
+				connected = false
 			}
 		}
 	}
```

Comment:
The above does the following:

1. make the close sync with the run() method.
2. Closing the client will only happen when run() execution is completed.
3. We check for `<- w.done` in two select, we do this to prioritize shutdown vs reading from queue on next execution loop, we also select on both the queue and the done channels and wait, first read will win here.